File: python/dec_06/dec_06.py
def try_squid_math(lines):
    # strip newlines and empty lines
    lines = [line.rstrip("\n") for line in lines]
    lines = [line for line in lines if line.strip()]
    # last line should be operations
    ops = lines[-1]
    # everything else should be numbers
    nums = lines[:-1]
    cols = list(zip(*nums, ops))
    problem_ranges = []
    # this is a problem range if any column has non-space digits
    # this is also bulshit!
    in_problem = False
    start = None
    for i, col in enumerate(cols):
        if all(c == ' ' for c in col):
            if in_problem:
                problem_ranges.append((start, i))
                in_problem = False
                start = None
        else:
            if not in_problem:
                in_problem = True
                start = i
    if in_problem:
        problem_ranges.append((start, len(cols)))
    answers = []
    for start, end in problem_ranges:
        op = ''.join(ops[start:end]).strip()
        # For each problem, for each number, collect one digit from each column (right to left) FML!!!!!
        num_digits = [[] for _ in range(end - start)]
        for col in range(end - 1, start - 1, -1):
            for row_idx in range(len(nums)):
                digit = nums[row_idx][col]
                if digit != ' ':
                    # please work this time
                    num_digits[end - 1 - col].append(digit)
        numbers = []
        for digits in num_digits:
            if digits:
                num_str = ''.join(digits)
                numbers.append(int(num_str))
        if op == '+':
            answers.append(sum(numbers))
        elif op == '*':
            prod = 1
            for n in numbers:
                prod *= n
            answers.append(prod)
        else:
            raise ValueError(f"somethings wrong again: {op}")
    return sum(answers)

with open("input_06.txt") as input:
    lines = [line.rstrip("\n") for line in input]

# Remove empty lines
lines = [line for line in lines if line.strip()]

print(f"answer: {try_squid_math(lines)}")

File: ruby/dec_06/dec_06.rb
#refactored as ruby this time
def try_cephalopod_math(lines)
  lines = lines.map(&:rstrip).reject { |line| line.strip.empty? }
  ops = lines[-1]
  nums = lines[0...-1]
  maxlen = ([ops.size] + nums.map(&:size)).max
  padded = nums.map { |row| row.ljust(maxlen) } + [ops.ljust(maxlen)]
  cols = padded.map(&:chars).transpose.map(&:join)
  problem_ranges = []
  in_problem = false
  start = nil
  cols.each_with_index do |col, i|
    if col.chars.all? { |c| c == ' ' }
      if in_problem
        problem_ranges << [start, i]
        in_problem = false
        start = nil
      end
    else
      unless in_problem
        in_problem = true
        start = i
      end
    end
  end
  problem_ranges << [start, cols.size] if in_problem
  answers = []
  problem_ranges.each do |start, end_idx|
    op = ops[start...end_idx].strip
    num_digits = Array.new(end_idx - start) { [] }
    (end_idx - 1).downto(start) do |col|
      nums.each_with_index do |row, row_idx|
        digit = row[col]
        num_digits[end_idx - 1 - col] << digit if digit != ' '
      end
    end
    numbers = num_digits.map { |digits| digits.empty? ? nil : digits.join.to_i }.compact
    if op == '+'
      answers << numbers.sum
    elsif op == '*'
      answers << numbers.inject(1) { |prod, n| prod * n }
    else
      raise "this broke again: #{op}"
    end
  end
  answers.sum
end

lines = File.readlines("input_06.txt").map(&:rstrip)
lines = lines.reject { |line| line.strip.empty? }
puts "answer: #{try_cephalopod_math(lines)}"
